packages/nimble-components/src/utilities/models/array-to-tree.ts

```diff
@@ -0,0 +1,121 @@
+// Modified from https://github.com/philipstanislaus/performant-array-to-tree/blob/v1.11.0/src/arrayToTree.ts
+// Copyright (c) 2022 philipstanislaus
+// SPDX-License-Identifier: MIT
+
+export interface TreeItem<Item> {
+    clientRecord: Item | undefined;
+    subRows: TreeItem<Item>[];
+    originalIndex: number | undefined;
+}
+
+export interface Config<Item> {
+    id: keyof Item;
+    parentId: keyof Item;
+}
+
+/**
+ * Unflattens an array to a tree with runtime O(n)
+ */
+export function arrayToTree<Item>(
+    items: readonly Item[],
+    config: Config<Item>
+): TreeItem<Item>[] {
+    const conf = config;
+
+    // the resulting unflattened tree
+    const rootItems: TreeItem<Item>[] = [];
+
+    // stores all already processed items with their ids as key so we can easily look them up
+    const lookup: { [id: string]: TreeItem<Item> } = {};
+
+    // stores all item ids that have not been added to the resulting unflattened tree yet
+    // this is an opt-in property, since it has a slight runtime overhead
+    const orphanIds: null | Set<string | number> = new Set();
+
+    // idea of this loop:
+    // whenever an item has a parent, but the parent is not yet in the lookup object, we store a preliminary parent
+    // in the lookup object and fill it with the data of the parent later
+    // if an item has no parentId, add it as a root element to rootItems
+    // [Nimble] Convert to a for-loop to have access to the item's index in the flat list
```

Comment:
Don't really need the `[Nimble]` marker here or below in other comments, it's all  nimble now :)